Implement motion detection similar to trips detection

Abyss777 opened the pull request:
Another try...
Implemented delayed motion events without timers, as you suggested.

`updateMotionState` looks complex only because I combined in one update (with new position) and correct (without position).

Positions-Events flow become not linear, because of that we need update events directly and some modifications.

More comments in code...

Abyss777 commented on src/org/traccar/database/ConnectionManager.java:
```diff
@@ -86,15 +88,18 @@ public void updateDevice(final long deviceId, String status, Date time) {
                     break;
             }
-            Event event = new Event(eventType, deviceId);
-            if (Context.getNotificationManager() != null) {
```
Do not need this check because switch is common for every events

Abyss777 commented on src/org/traccar/database/ConnectionManager.java:
```diff
@@ -140,10 +145,14 @@ public synchronized void updatePosition(Position position) {
+        }
+    }
+
+    public synchronized void updatePosition(long userId, Position position) {
```
Implemented function to update position for single user

Abyss777 commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -66,6 +76,11 @@ public DeviceManager(DataManager dataManager) {
         lookupGroupsAttribute = config.getBoolean("deviceManager.lookupGroupsAttribute");
         fallbackToText = config.getBoolean("command.fallbackToSms");
+
+        minimalParkingDuration = config.getLong("report.trip.minimalParkingDuration", 300) * 1000;
```
Constants initialized with same values as for trips/stops

Abyss777 commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -387,4 +402,54 @@ public void sendCommand(Command command) throws Exception {
+                }
+                if (newMotion) {
+                    if (potentialTime + minimalTripDuration <= currentTime
+                            || distance == null || distance >= minimalTripDistance) {
```
Detect motion by time and by distance

Abyss777 commented on src/org/traccar/database/NotificationManager.java:
```diff
@@ -62,6 +66,9 @@ public void updateEvent(Event event, Position position) {
                     if (notification.getWeb()) {
                         Context.getConnectionManager().updateEvent(userId, event);
+                        if (withPosition) {
+                            Context.getConnectionManager().updatePosition(userId, position);
```
Because related position from past we need resent it after event to correctly link in web-client in "Events panel"

tananaev replied:
Please elaborate. It doesn't seem right to send position twice. Am I understanding correctly that it's what you are doing?

Abyss777 commented on src/org/traccar/events/MotionEventHandler.java:
```diff
@@ -37,19 +36,7 @@
-            return Collections.singleton(
-                    new Event(Event.TYPE_DEVICE_STOPPED, position.getDeviceId(), position.getId()));
-        }
+        Context.getDeviceManager().updateMotionState(position.getDeviceId(), position);
```
Return nothing because event never be related to current position.

tananaev replied:
I think we should move motion detection logic back. See comment above.

tananaev commented on src/org/traccar/database/ConnectionManager.java:
```diff
@@ -41,6 +41,7 @@
 
     private final long deviceTimeout;
     private final boolean enableStatusEvents;
+    private final boolean correctMotionState;
```
What does this parameter do? I think we need a better name because it's not very clear.

Abyss777 replied:
It enables additional check motion state on offline/unknown status changes.

tananaev replied:
Can't think of any good name at the moment, we definitely need something better here.

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -52,6 +55,13 @@
     private final Map<Long, Position> positions = new ConcurrentHashMap<>();
 
+    private final Map<Long, Boolean> motionStates = new ConcurrentHashMap<>();
+    private final Map<Long, Position> motionStatePositions = new ConcurrentHashMap<>();
```
Don't we already store positions above? or is it a first position? In that case do we really need to store the whole position or just the time would be enough?

Abyss777 replied:
We need also distance for detect moving, link `positionId` to new event and pass appropriate `position` (whole) to notification.

tananaev replied:
Makes sense. Ignore my comment.

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -52,6 +55,13 @@
 
     private final Map<Long, Position> positions = new ConcurrentHashMap<>();
 
+    private final Map<Long, Boolean> motionStates = new ConcurrentHashMap<>();
```
Do you think it's a good idea to create some "device state" class to store last position, motion status, and possibly some other information in future?

Abyss777 replied:
Not sure, may be it will be more convenient.

tananaev replied:
It is faster than queries each parameter separate and it will enable us to easily add some extra information later.

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -387,4 +402,54 @@ public void sendCommand(Command command) throws Exception {
         return result;
     }
+
+    public void updateMotionState(long deviceId, Position position) {
```
It would be nice to extract logic back into motionhandler to avoid adding to much code device manager.

Abyss777 replied:
We can't `MotionHandler` are multiple for every pipeline, but we need some singleton to store per device motion states. Or am I mistaken?

tananaev replied:
Yes, my idea was that you store any required state in device manager, but actual calculation code is here. This way we separate the code to make it easier to read.

tananaev commented on src/org/traccar/database/NotificationManager.java:
```diff
@@ -48,6 +48,10 @@ public NotificationManager(DataManager dataManager) {
     }
 
     public void updateEvent(Event event, Position position) {
+        updateEvent(event, position, false);
```
Is there any time when position is not null and we don't want to update it? Possibly not for this pull request, but it would help with handling events in the web app if event and position come in the same message.

Abyss777 replied:
Positions come through pipeline and go to web-client at the end.

tananaev commented on test/org/traccar/events/MotionEventHandlerTest.java:
```diff
@@ -1,29 +0,0 @@
-import org.traccar.model.Event;
-import org.traccar.model.Position;
-
-public class MotionEventHandlerTest extends BaseTest {
```
It would be nice to implement some unit testing for the new logic.

Abyss777 commented:
I'm not sure how to implement such events without resend position. For "Events panel" it should come after Event.

Abyss777 commented:
Drew flow scheme to represent it visually.
![flows](https://user-images.githubusercontent.com/5688080/29065503-bf948d0a-7c45-11e7-873c-d7d9d08958ab.png)

I feel we need implement something essential (base) for such Events. I'm sure we want make some other events similar.

Abyss777 commented:
One option is We might not resend past position but request it via API if user click it on Events Panel

tananaev commented:
Requesting position through API is also not ideal solution. We have the position. Why would we need to duplicate effort.

I think maybe we should modify pipeline to push events up the chain to the main event handler. This way we can keep position and event together. This will require quite a bit of refactoring, but seems like the right solution. It will also unify the way events and positions are pushed to the web app.

Abyss777 commented:
But still, do not understand how it help. Event still be related to position from past (user can login in period between this position and related event). Also we have some event types not related to positions and still need a way to generate and transport events directly (without pipeline)
Do not think splitting a ways how we deliver events is good idea (through pipeline or directly)

tsmgeek commented:
Is this for devices that do not have movement detection built in or are you defining movement as a positional change?

tananaev commented:
I see the problem now. It's an old position, not current one. I missed that point. Then using API might be a good idea.

Abyss777 commented on src/org/traccar/model/DeviceState.java:
```diff
@@ -0,0 +1,41 @@
+
+public class DeviceState {
+
+    private Boolean motionState;
```
It is not primitive intentionally. To determine if it is not initialized.

Abyss777 commented on src/org/traccar/database/ConnectionManager.java:
```diff
@@ -49,6 +53,9 @@
     public ConnectionManager() {
         deviceTimeout = Context.getConfig().getLong("status.timeout", DEFAULT_TIMEOUT) * 1000;
         enableStatusEvents = Context.getConfig().getBoolean("event.enable");
+        if (Context.getConfig().getBoolean("status.updateDeviceState")) {
```
Is the name OK?

tananaev replied:
I think it's OK. I'm assuming that the idea is that state will have more information than just motion state.

Abyss777 commented:
Replaced the commit.
Implemented `DeviceState` class, in future we can use it for overspeeding state or other events.
Events now return back in flow, not send directly.
Implemented test.

I decided not over complicate `MotionEventHandler.updateMotionState` to not use it if we handle offline/unknown status. Just implemented simple variant in `ConnectionManager`

Also fixed issue in `ReportUtils`, parameters were initialized in wrong order.

tananaev commented on src/org/traccar/model/DeviceState.java:
```diff
@@ -0,0 +1,41 @@
+ */
+package org.traccar.model;
+
+public class DeviceState {
```
Do you think it makes sense to add last position here as well?

Abyss777 replied:
Yes I think it is reasonable. Concentrate it in one place.
Also I was thinking that it is good place for device Status, but remember that we notify web client about it.

tananaev replied:
I guess we can leave it as it is for now...

tananaev commented on src/org/traccar/database/ConnectionManager.java:
```diff
@@ -126,6 +142,29 @@ public void run(Timeout timeout) throws Exception {
+        }
+        Event result = null;
+        Boolean oldMotion = deviceState.getMotionState();
+        long currentTime = new Date().getTime();
```
Use `System.currentTimeMillis()` instead.

tananaev commented on src/org/traccar/events/MotionEventHandler.java:
```diff
@@ -21,11 +21,60 @@
+    private TripsConfig tripsConfig;
+
+    public MotionEventHandler() {
+        if (Context.getConfig() != null) {
```
Is there any case when we don't have config?

Abyss777 replied:
Forgot to remove this check, when make next function `static` for tests.

tananaev commented on src/org/traccar/database/ConnectionManager.java:
```diff
@@ -126,6 +142,29 @@ public void run(Timeout timeout) throws Exception {
+        Boolean oldMotion = deviceState.getMotionState();
+        long currentTime = new Date().getTime();
+        boolean newMotion = !oldMotion;
+        Position potentialPosition = deviceState.getMotionPosition();
```
I think we should just name this `motionPosition`.

tananaev commented on src/org/traccar/database/ConnectionManager.java:
```diff
@@ -126,6 +142,29 @@ public void run(Timeout timeout) throws Exception {
+        boolean newMotion = !oldMotion;
+        Position potentialPosition = deviceState.getMotionPosition();
+        if (potentialPosition != null) {
+            long potentialTime = potentialPosition.getFixTime().getTime()
```
I don't like `potential` prefix. Maybe `thresholdTime`?

Abyss777 commented:
Renamed variables by next logic `potentialPosition` to `motionPosition`, `potentialTime` to `motionPositionTime`, but it is too long, leaved just `motionTime`

Abyss777 commented:
Sorry, occasionally closed.

tananaev commented:
Merged, thanks.

Abyss777 commented:
I applied this and next changes to my server at Friday and tested them on real data during weekend.
It works correctly with some parameters tuning, but wants a bit more.
We have second parameter to detect motion - distance, but we detect stop only by time. I believe we should add second parameter to detect stop.
I'm suggesting - Ignition (of course if position contains it). If ignition is ON - it does not mean that car is moving, but usually car can't move with stopped engine. What do you think?

tananaev commented:
Makes sense to me.
